core/drand_beacon.go

```diff
@@ -283,8 +283,21 @@ func (bp *BeaconProcess) transition(oldGroup *key.Group, oldPresent, newPresent
 
 // Stop simply stops all drand operations.
 func (bp *BeaconProcess) Stop(ctx context.Context) {
+	select {
+	case <-bp.exitCh:
+		bp.log.Errorw("Trying to stop an already stopping beacon process", "id", bp.getBeaconID())
+		return
+	default:
+		bp.log.Debugw("Stopping BeaconProcess", "id", bp.getBeaconID())
+	}
 	bp.StopBeacon()
-	bp.exitCh <- true
+	// we wait until we can send on the channel or the context got canceled
+	select {
+	case bp.exitCh <- true:
```

Comment:
This took me a while to grok how it was different from the above select x( does it strictly need to be two `select`s or could it all be in one?

Reply:
+1 on this being a bit harder to read than it should be.

I think I might change how this works for the work on restarting a beacon without restarting the node.